lib/Target/HSACO/HSACOTranslation.cpp

```diff
@@ -0,0 +1,187 @@
+#include "llvm/Transforms/Utils/Cloning.h"
+#include <filesystem>
+#include <iostream>
+#include <memory>
+#include <random>
+
+namespace {
+
+void init_llvm() {
+  LLVMInitializeAMDGPUTarget();
+  LLVMInitializeAMDGPUTargetInfo();
+  LLVMInitializeAMDGPUTargetMC();
+  LLVMInitializeAMDGPUAsmParser();
+  LLVMInitializeAMDGPUAsmPrinter();
+}
+
+std::mt19937_64 *InitRngWithRandomSeed() {
+  std::random_device device("/dev/urandom");
+  return new std::mt19937_64(device());
+}
+
+uint64_t New64() {
+  static std::mt19937_64 *rng = InitRngWithRandomSeed();
+  return (*rng)();
+}
+
+std::unique_ptr<llvm::TargetMachine>
+initialize_module(llvm::Module *module, const std::string &triple,
+                  const std::string &proc, const std::string &features) {
+  // verify and store llvm
+  llvm::legacy::PassManager pm;
+  pm.add(llvm::createVerifierPass());
+  pm.run(*module);
+
+  module->setTargetTriple(triple);
+
+  std::string error;
+  auto target =
+      llvm::TargetRegistry::lookupTarget(module->getTargetTriple(), error);
+  llvm::TargetOptions opt;
+  opt.AllowFPOpFusion = llvm::FPOpFusion::Fast;
+  opt.UnsafeFPMath = false;
+  opt.NoInfsFPMath = false;
+  opt.NoNaNsFPMath = true;
+  llvm::TargetMachine *machine = target->createTargetMachine(
+      module->getTargetTriple(), proc, features, opt, llvm::Reloc::PIC_,
+      std::nullopt, llvm::CodeGenOpt::Aggressive);
+
+  module->setDataLayout(machine->createDataLayout());
+
+  for (llvm::Function &f : module->functions())
+    f.addFnAttr(llvm::Attribute::AlwaysInline);
+
+  return std::unique_ptr<llvm::TargetMachine>(machine);
+}
+
+std::string generate_amdgcn_assembly(llvm::Module *module,
+                                     const std::string &triple,
+                                     const std::string &proc,
+                                     const std::string &features) {
+  auto machine = initialize_module(module, triple, proc, features);
+  llvm::SmallVector<char, 0> buffer;
+  llvm::legacy::PassManager pass;
+  llvm::raw_svector_ostream stream(buffer);
+
+  // emit
+  machine->addPassesToEmitFile(pass, stream, nullptr,
+                               llvm::CodeGenFileType::CGFT_AssemblyFile);
+  pass.run(*module);
+
+  std::string amdgcn(buffer.begin(), buffer.end());
+  if (::triton::tools::getBoolEnv("AMDGCN_ENABLE_DUMP")) {
+    std::cout << "// -----// AMDGCN Dump //----- //\n" << amdgcn << std::endl;
+  }
+
+  return amdgcn;
+}
+
+std::string generate_hsaco(llvm::Module *module, const std::string &triple,
+                           const std::string &proc,
+                           const std::string &features) {
+  auto machine = initialize_module(module, triple, proc, features);
+
+  // create dir for binary and hsaco
+  std::error_code ec;
+  std::string kernel_name = std::to_string(New64());
+  llvm::SmallString<256> kernel_dir_unique;
+  ec = llvm::sys::fs::createUniqueDirectory(std::filesystem::temp_directory_path().string(), kernel_dir_unique);
```

Comment:
Now that the risk of conflict is solved with this directory, would it make sense to simplify the name of the kernel?
At least the kernel file name could just be always the same.
But perhaps a different kernel name to have a cleaner name for a better debug and performance user experience?
`amd_triton_0`, `amd_triton_1`, etc. or whatever. So just using a static counter instead of a random number generator. Just a guess, since I do not understand where the function name comes from.

Reply:
using the unique dir as the kernel name.